File: apps/gateway/src/tmux-client/lifecycle-emitter.ts
import type { Device, EventType, SiteSettings, TmuxWindow, WebhookEvent } from '@tmex/shared';
import { getSiteSettings } from '../db';
import type { LifecycleEventEmitter } from './connection-types';
import { diffSnapshotClosures } from './snapshot-diff';

export interface LifecycleEmitterContext {
  getDevice: () => Device | null;
  getSessionName: () => string;
  isEmittable: () => boolean;
  getSnapshotWindows: () => ReadonlyMap<string, TmuxWindow>;
  notifyEvent?: LifecycleEventEmitter;
  settingsProvider?: () => SiteSettings;
  resolveCustomName?: (kind: 'window' | 'pane', nativeId: string) => string | undefined;
}

// local/ssh 两连接类共享的生命周期事件发射器。发射是旁路观测，任何一步失败
//（settings 读取、事件回调）都不允许影响连接主控制流，整体兜底只记日志。
export class ConnectionLifecycleEmitter {
  private sessionClosedEmittedFlag = false;

  constructor(private readonly ctx: LifecycleEmitterContext) {}

  reset(): void {
    this.sessionClosedEmittedFlag = false;
  }

  // 本次连接是否已因 session gone 发出 session_closed。断开告警桥据此
  // 抑制同一物理事件的 device_disconnect，避免双发。
  get sessionClosedEmitted(): boolean {
    return this.sessionClosedEmittedFlag;
  }

  emit(eventType: EventType, tmux: WebhookEvent['tmux'], payload?: Record<string, unknown>): void {
    const notifyEvent = this.ctx.notifyEvent;
    if (!notifyEvent) {
      return;
    }
    try {
      const device = this.ctx.getDevice();
      if (!device) {
        return;
      }
      const settings = (this.ctx.settingsProvider ?? getSiteSettings)();
      notifyEvent(eventType, {
        site: { name: settings.siteName, url: settings.siteUrl },
        device: { id: device.id, name: device.name, type: device.type, host: device.host },
        tmux,
        payload,
      });
    } catch (err) {
      console.error(`[tmux-client] lifecycle event emit failed (${eventType}):`, err);
    }
  }

  notifySessionClosed(message: string): void {
    if (this.sessionClosedEmittedFlag) {
      return;
    }
    this.sessionClosedEmittedFlag = true;
    this.emit(
      'session_closed',
      { sessionName: this.ctx.getSessionName() },
      { message: message.split(/\r?\n/)[0]?.trim() }
    );
  }

  notifySessionCreated(): void {
    this.emit('session_created', { sessionName: this.ctx.getSessionName() });
  }

  // 快照 diff 产生 window/pane 关闭事件。首帧（prev 为空）、无效快照（next 为空）与
  // 断开路径一律跳过，避免误报。
  emitSnapshotClosures(prev: ReadonlyMap<string, TmuxWindow>): void {
    const next = this.ctx.getSnapshotWindows();
    if (prev.size === 0 || next.size === 0 || !this.ctx.isEmittable()) {
      return;
    }
    const sessionName = this.ctx.getSessionName();
    const { closedWindows, closedPanes } = diffSnapshotClosures(prev, next);
    const windowDisplayName = (window: TmuxWindow) =>
      this.ctx.resolveCustomName?.('window', window.id) ?? window.name;
    for (const window of closedWindows) {
      this.emit(
        'tmux_window_close',
        { sessionName, windowId: window.id, windowIndex: window.index },
        { windowName: windowDisplayName(window) }
      );
    }
    for (const { pane, window } of closedPanes) {
      this.emit(
        'tmux_pane_close',
        {
          sessionName,
          windowId: window.id,
          windowIndex: window.index,
          paneId: pane.id,
          paneIndex: pane.index,
          // 快照不含改名 overlay，从 ctx 注入的 projection 查询取（用户改名优先）
          paneTitle: this.ctx.resolveCustomName?.('pane', pane.id) ?? pane.title,
          paneCurrentCommand: pane.currentCommand,
          paneCurrentPath: pane.currentPath,
        },
        { windowName: windowDisplayName(window) }
      );
    }
  }
}

File: apps/gateway/src-rust/tmux/lifecycle_emitter.rs
use std::collections::BTreeMap;

use tmex_protocol::WindowWire;

use super::metadata_projection::ProjectionEntityKind;
use super::snapshot_diff::diff_snapshot_closures;

#[derive(Clone, Copy, Debug, PartialEq, Eq)]
pub enum LifecycleEventKind {
    SessionCreated,
    SessionClosed,
    TmuxWindowClose,
    TmuxPaneClose,
}

#[derive(Clone, Debug, Default, PartialEq, Eq)]
pub struct LifecycleTmuxContext {
    pub session_name: Option<String>,
    pub window_id: Option<String>,
    pub window_index: Option<u16>,
    pub pane_id: Option<String>,
    pub pane_index: Option<u16>,
    pub pane_title: Option<String>,
    pub pane_current_command: Option<String>,
    pub pane_current_path: Option<String>,
}

#[derive(Clone, Debug, PartialEq, Eq)]
pub struct LifecycleEvent {
    pub kind: LifecycleEventKind,
    pub tmux: LifecycleTmuxContext,
    pub payload: BTreeMap<String, String>,
}

pub trait TmuxLifecycleSink: Send + Sync + 'static {
    fn publish(&self, device_id: String, event: LifecycleEvent);
}

impl<Publish> TmuxLifecycleSink for Publish
where
    Publish: Fn(String, LifecycleEvent) + Send + Sync + 'static,
{
    fn publish(&self, device_id: String, event: LifecycleEvent) {
        self(device_id, event);
    }
}

#[derive(Clone, Debug, Default, PartialEq, Eq)]
pub struct ConnectionLifecycleEmitter {
    session_closed_emitted: bool,
}

impl ConnectionLifecycleEmitter {
    pub fn new() -> Self {
        Self::default()
    }

    pub fn reset(&mut self) {
        self.session_closed_emitted = false;
    }

    pub fn session_closed_emitted(&self) -> bool {
        self.session_closed_emitted
    }

    pub fn notify_session_closed(
        &mut self,
        session_name: impl Into<String>,
        message: &str,
    ) -> Option<LifecycleEvent> {
        if self.session_closed_emitted {
            return None;
        }
        self.session_closed_emitted = true;
        let first_line = message.lines().next().unwrap_or_default().trim().to_owned();
        Some(LifecycleEvent {
            kind: LifecycleEventKind::SessionClosed,
            tmux: LifecycleTmuxContext {
                session_name: Some(session_name.into()),
                ..LifecycleTmuxContext::default()
            },
            payload: BTreeMap::from([("message".to_owned(), first_line)]),
        })
    }

    pub fn notify_session_created(session_name: impl Into<String>) -> LifecycleEvent {
        LifecycleEvent {
            kind: LifecycleEventKind::SessionCreated,
            tmux: LifecycleTmuxContext {
                session_name: Some(session_name.into()),
                ..LifecycleTmuxContext::default()
            },
            payload: BTreeMap::new(),
        }
    }

    pub fn snapshot_closures<ResolveName>(
        &self,
        previous: &BTreeMap<String, WindowWire>,
        next: &BTreeMap<String, WindowWire>,
        emittable: bool,
        session_name: &str,
        mut resolve_custom_name: ResolveName,
    ) -> Vec<LifecycleEvent>
    where
        ResolveName: FnMut(ProjectionEntityKind, &str) -> Option<String>,
    {
        if previous.is_empty() || next.is_empty() || !emittable {
            return Vec::new();
        }
        let closures = diff_snapshot_closures(previous, next);
        let mut events =
            Vec::with_capacity(closures.closed_windows.len() + closures.closed_panes.len());
        for window in closures.closed_windows {
            let window_name = resolve_custom_name(ProjectionEntityKind::Window, &window.id)
                .unwrap_or_else(|| window.name.clone());
            events.push(LifecycleEvent {
                kind: LifecycleEventKind::TmuxWindowClose,
                tmux: LifecycleTmuxContext {
                    session_name: Some(session_name.to_owned()),
                    window_id: Some(window.id),
                    window_index: Some(window.index),
                    ..LifecycleTmuxContext::default()
                },
                payload: BTreeMap::from([("windowName".to_owned(), window_name)]),
            });
        }
        for closed in closures.closed_panes {
            let pane_title = resolve_custom_name(ProjectionEntityKind::Pane, &closed.pane.id)
                .or(closed.pane.title.clone());
            let window_name = resolve_custom_name(ProjectionEntityKind::Window, &closed.window.id)
                .unwrap_or_else(|| closed.window.name.clone());
            events.push(LifecycleEvent {
                kind: LifecycleEventKind::TmuxPaneClose,
                tmux: LifecycleTmuxContext {
                    session_name: Some(session_name.to_owned()),
                    window_id: Some(closed.window.id),
                    window_index: Some(closed.window.index),
                    pane_id: Some(closed.pane.id),
                    pane_index: Some(closed.pane.index),
                    pane_title,
                    pane_current_command: closed.pane.current_command,
                    pane_current_path: closed.pane.current_path,
                },
                payload: BTreeMap::from([("windowName".to_owned(), window_name)]),
            });
        }
        events
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn session_close_is_once_per_connection_and_reset_rearms_it() {
        let mut emitter = ConnectionLifecycleEmitter::new();
        let first = emitter.notify_session_closed("work", "line one\nline two");
        assert_eq!(
            first.unwrap().payload.get("message").map(String::as_str),
            Some("line one")
        );
        assert!(emitter.notify_session_closed("work", "again").is_none());
        emitter.reset();
        assert!(emitter.notify_session_closed("work", "after").is_some());
    }
}
